File: Tests/skkserv-compoundTests/SKKServerTests.swift
// SPDX-License-Identifier: MIT

import Testing
import Foundation
import Logging
import NIOCore
@testable import skkserv_compound

@Suite("SKKServer response")
struct SKKServerResponseTests {
    private static let testLogger = Logger(label: "server-test")

    private func makeServer(
        user: [(String, [String])] = [],
        system: [(String, [String])] = [],
        config: CompoundGeneratorConfig = CompoundGeneratorConfig()
    ) -> SKKServer {
        let snapshot = DictionaryLoader.buildSnapshot(
            user: user.map { ParsedEntry(reading: $0.0, candidates: $0.1) },
            system: system.map { ParsedEntry(reading: $0.0, candidates: $0.1) }
        )
        let store = DictionaryStore(initial: snapshot)
        return SKKServer(
            version: "test",
            logger: Self.testLogger,
            store: store,
            generatorConfig: config
        )
    }

    @Test("opcode 1: 候補ありで 1/.../\\n を返す")
    func opcode1ReturnsCandidates() async {
        let server = makeServer(system: [
            ("せいそう", ["清掃"]),
            ("ぎょうしゃ", ["業者"])
        ])
        let response = await server.candidateResponse(for: "せいそうぎょうしゃ ")
        #expect(response == "1/清掃業者/\n")
    }

    @Test("opcode 1: 候補なしで 4\\n を返す")
    func opcode1ReturnsFourOnMiss() async {
        let server = makeServer()
        let response = await server.candidateResponse(for: "そんざいしない ")
        #expect(response == "4\n")
    }

    @Test("入力読みの前後空白・改行を除去する")
    func trimsWhitespace() async {
        let server = makeServer(system: [
            ("あ", ["亜"]),
            ("い", ["胃"])
        ])
        let response = await server.candidateResponse(for: "  あい\n")
        #expect(response == "1/亜胃/\n")
    }

    @Test("末尾の送り仮名マーカーを除去する (送りあり辞書未登録なら okuri-ari 連結は出ない)")
    func stripsOkuriMarker() async {
        // okuri-ari 辞書に該当エントリがない場合、 okuriPrefix 付きリクエストの応答は
        // 送りあり連結のみで構成されるため空 (送りなし結果は混ぜない)。
        let server = makeServer(system: [
            ("あ", ["亜"]),
            ("い", ["胃"])
        ])
        let response = await server.candidateResponse(for: "あいi")
        #expect(response == "4\n")
    }

    @Test("opcode 1: もんだいなs で 1/問題無/問題済/\\n を返す")
    func opcode1OkuriAriCompound() async {
        let user = [
            ParsedEntry(reading: "もんだい", candidates: ["問題"], isOkuriAri: false),
            ParsedEntry(reading: "なs", candidates: ["無", "済"], isOkuriAri: true)
        ]
        let snapshot = DictionaryLoader.buildSnapshot(user: user, system: [])
        let server = SKKServer(
            version: "test",
            logger: Self.testLogger,
            store: DictionaryStore(initial: snapshot)
        )
        let response = await server.candidateResponse(for: "もんだいなs ")
        #expect(response == "1/問題無/問題済/\n")
    }

    @Test("opcode 1: 全 ASCII 入力 kawaii では okuriPrefix を抽出せず abbrev 連結を引く")
    func opcode1AsciiInputDoesNotExtractOkuri() async {
        let server = makeServer(system: [
            ("ka", ["カ"]),
            ("waii", ["ワイイ"])
        ])
        let response = await server.candidateResponse(for: "kawaii ")
        // sanitize は okuriPrefix を抽出せず body=`kawaii` のまま abbrev 連結探索する。
        #expect(response == "1/カワイイ/\n")
    }
}

@Suite("SKKServer opcodes")
struct SKKServerOpcodeTests {
    private static let testLogger = Logger(label: "server-opcode-test")

    private func makeServer(
        user: [(String, [String])] = [],
        system: [(String, [String])] = []
    ) -> SKKServer {
        let snapshot = DictionaryLoader.buildSnapshot(
            user: user.map { ParsedEntry(reading: $0.0, candidates: $0.1) },
            system: system.map { ParsedEntry(reading: $0.0, candidates: $0.1) }
        )
        return SKKServer(
            version: "v1",
            serverName: "skkserv-test",
            logger: Self.testLogger,
            store: DictionaryStore(initial: snapshot)
        )
    }

    @Test("opcode 0 はコネクションを閉じる")
    func opcodeZeroCloses() async {
        let server = makeServer()
        #expect(await server.handleOpcode("0", operand: "", host: "127.0.0.1", port: 1178) == .close)
    }

    @Test("opcode 2 で version 応答を返す")
    func opcodeTwoReturnsVersion() async {
        let server = makeServer()
        #expect(await server.handleOpcode("2", operand: "", host: "127.0.0.1", port: 1178) == .reply("skkserv-test/v1 "))
    }

    @Test("opcode 3 で host/port 応答を返す")
    func opcodeThreeReturnsHostPort() async {
        let server = makeServer()
        let result = await server.handleOpcode("3", operand: "", host: "127.0.0.1", port: 1178)
        guard case .reply(let body) = result else {
            Issue.record("expected reply, got \(result)")
            return
        }
        #expect(body.hasSuffix("/127.0.0.1:1178 "))
    }

    @Test("opcode 4 は常に 4\\n を返す")
    func opcodeFourReturnsFour() async {
        let server = makeServer()
        #expect(await server.handleOpcode("4", operand: "なにか ", host: "127.0.0.1", port: 1178) == .reply("4\n"))
    }

    @Test("opcode 1 はパイプライン経由でも 1/.../\\n を返す")
    func opcodeOneIntegrates() async {
        let server = makeServer(system: [
            ("あ", ["亜"]),
            ("い", ["胃"])
        ])
        #expect(await server.handleOpcode("1", operand: "あい ", host: "127.0.0.1", port: 1178) == .reply("1/亜胃/\n"))
    }

    @Test("未対応 opcode は ignore")
    func unsupportedOpcodeIgnored() async {
        let server = makeServer()
        #expect(await server.handleOpcode("9", operand: "", host: "127.0.0.1", port: 1178) == .ignore)
    }
}

@Suite("Incoming charset")
struct IncomingCharsetTests {
    @Test("EUC-JP のバイト列を読みとして復号できる")
    func decodesEucJp() {
        // "1あい " encoded as EUC-JP
        let text = "1あい "
        guard let bytes = text.data(using: .japaneseEUC) else {
            Issue.record("EUC-JP encoding failed")
            return
        }
        let roundTrip = String(data: bytes, encoding: .japaneseEUC)
        #expect(roundTrip == text)
        // Sanitize the operand portion the way the server does.
        let operand = String(text.dropFirst())
        let (body, okuri) = sanitizeYomi(operand)
        #expect(body == "あい")
        #expect(okuri == nil)
    }
}

@Suite("SKKServer.extractMessages")
struct ExtractMessagesTests {
    private static let testLogger = Logger(label: "extract-test")

    private func makeBuffer(_ bytes: [UInt8]) -> ByteBuffer {
        var buffer = ByteBufferAllocator().buffer(capacity: bytes.count)
        buffer.writeBytes(bytes)
        return buffer
    }

    private func makeBuffer(_ string: String, encoding: String.Encoding = .utf8) -> ByteBuffer {
        let data = string.data(using: encoding) ?? Data()
        return makeBuffer(Array(data))
    }

    @Test("1 つの ByteBuffer に複数のリクエストが詰まっていても全て切り出せる")
    func extractsMultipleRequestsFromSingleBuffer() {
        var buffer = makeBuffer("1あい 1うえ ")
        let messages = SKKServer.extractMessages(
            buffer: &buffer,
            charset: .utf8,
            logger: Self.testLogger
        )
        #expect(messages.count == 2)
        #expect(messages[0].opcode == "1")
        #expect(messages[0].operand == "あい")
        #expect(messages[1].opcode == "1")
        #expect(messages[1].operand == "うえ")
        #expect(buffer.readableBytes == 0)
    }

    @Test("リクエストが複数の ByteBuffer に分断されて届いても結合できる")
    func reassemblesAcrossBuffers() {
        let full = "1あい "
        let fullBytes = Array(full.data(using: .utf8)!)
        let splitAt = fullBytes.count / 2
        var buffer = makeBuffer(Array(fullBytes[..<splitAt]))

        let firstPass = SKKServer.extractMessages(
            buffer: &buffer,
            charset: .utf8,
            logger: Self.testLogger
        )
        #expect(firstPass.isEmpty)
        #expect(buffer.readableBytes > 0)

        var second = makeBuffer(Array(fullBytes[splitAt...]))
        buffer.writeBuffer(&second)
        let secondPass = SKKServer.extractMessages(
            buffer: &buffer,
            charset: .utf8,
            logger: Self.testLogger
        )
        #expect(secondPass.count == 1)
        #expect(secondPass[0].opcode == "1")
        #expect(secondPass[0].operand == "あい")
        #expect(buffer.readableBytes == 0)
    }

    @Test("LF 区切りも space 区切りも両方認識する")
    func recognizesBothDelimiters() {
        var buffer = makeBuffer("1あ\n1い ")
        let messages = SKKServer.extractMessages(
            buffer: &buffer,
            charset: .utf8,
            logger: Self.testLogger
        )
        #expect(messages.count == 2)
        #expect(messages[0].opcode == "1")
        #expect(messages[0].operand == "あ")
        #expect(messages[1].opcode == "1")
        #expect(messages[1].operand == "い")
    }

    @Test("区切り未到達の末尾はバッファに残す")
    func keepsUnterminatedTail() {
        var buffer = makeBuffer("1あい 1うえ")
        let messages = SKKServer.extractMessages(
            buffer: &buffer,
            charset: .utf8,
            logger: Self.testLogger
        )
        #expect(messages.count == 1)
        #expect(messages[0].operand == "あい")
        #expect(buffer.readableBytes > 0)
    }

    @Test("EUC-JP のバイト列も同じ境界規則で切り出せる")
    func extractsEucJpRequests() {
        var buffer = makeBuffer("1あい 1うえ ", encoding: .japaneseEUC)
        let messages = SKKServer.extractMessages(
            buffer: &buffer,
            charset: .japaneseEUC,
            logger: Self.testLogger
        )
        #expect(messages.count == 2)
        #expect(messages[0].operand == "あい")
        #expect(messages[1].operand == "うえ")
    }

    @Test("空ペイロード (連続区切り) はスキップする")
    func skipsEmptyPayloads() {
        var buffer = makeBuffer("  1あ ")
        let messages = SKKServer.extractMessages(
            buffer: &buffer,
            charset: .utf8,
            logger: Self.testLogger
        )
        #expect(messages.count == 1)
        #expect(messages[0].opcode == "1")
        #expect(messages[0].operand == "あ")
    }
}

@Suite("sanitizeYomi")
struct SanitizeYomiTests {
    @Test("trim する; okuriPrefix なし")
    func trimsAndStrips() {
        let a = sanitizeYomi(" あい \n")
        #expect(a.body == "あい")
        #expect(a.okuriPrefix == nil)

        // 全 ASCII 入力は abbrev とみなし okuriPrefix を抽出しない
        let abc = sanitizeYomi("abc")
        #expect(abc.body == "abc")
        #expect(abc.okuriPrefix == nil)
    }

    @Test("ひらがな + ASCII 1 文字なら okuriPrefix として保持する")
    func extractsOkuriPrefix() {
        let s = sanitizeYomi("おくr")
        #expect(s.body == "おく")
        #expect(s.okuriPrefix == "r")

        let m = sanitizeYomi("もんだいなs ")
        #expect(m.body == "もんだいな")
        #expect(m.okuriPrefix == "s")
    }

    @Test("空入力は (\"\", nil) を返す")
    func returnsEmptyTuple() {
        let s = sanitizeYomi("   \n")
        #expect(s.body == "")
        #expect(s.okuriPrefix == nil)
    }
}

File: Tests/skkserver_tests.rs
// SPDX-License-Identifier: MIT

use skkserv_compound::dictionary::ParsedEntry;
use skkserv_compound::generator::CompoundGeneratorConfig;
use skkserv_compound::loader::build_snapshot;
use skkserv_compound::server::{
    IncomingCharset, OpcodeResult, SkkServer, extract_messages, sanitize_yomi,
};
use skkserv_compound::store::DictionaryStore;

fn nashi(reading: &str, candidates: &[&str]) -> ParsedEntry {
    ParsedEntry::new(
        reading.to_string(),
        candidates.iter().map(|s| s.to_string()).collect(),
    )
}

fn ari(reading: &str, candidates: &[&str]) -> ParsedEntry {
    ParsedEntry::with_okuri(
        reading.to_string(),
        candidates.iter().map(|s| s.to_string()).collect(),
        true,
    )
}

fn make_server(user: &[(&str, &[&str])], system: &[(&str, &[&str])]) -> SkkServer {
    let u: Vec<ParsedEntry> = user.iter().map(|(r, c)| nashi(r, c)).collect();
    let s: Vec<ParsedEntry> = system.iter().map(|(r, c)| nashi(r, c)).collect();
    let snap = build_snapshot(&u, &s);
    SkkServer::new(
        "test",
        "skkserv-compound",
        DictionaryStore::with_initial(snap),
        CompoundGeneratorConfig::default(),
    )
}

#[tokio::test]
async fn opcode1_returns_candidates() {
    let server = make_server(&[], &[("せいそう", &["清掃"]), ("ぎょうしゃ", &["業者"])]);
    let response = server.candidate_response("せいそうぎょうしゃ ").await;
    assert_eq!(response, "1/清掃業者/\n");
}

#[tokio::test]
async fn opcode1_returns_four_on_miss() {
    let server = make_server(&[], &[]);
    let response = server.candidate_response("そんざいしない ").await;
    assert_eq!(response, "4\n");
}

#[tokio::test]
async fn trims_whitespace() {
    let server = make_server(&[], &[("あ", &["亜"]), ("い", &["胃"])]);
    let response = server.candidate_response("  あい\n").await;
    assert_eq!(response, "1/亜胃/\n");
}

#[tokio::test]
async fn strips_okuri_marker() {
    let server = make_server(&[], &[("あ", &["亜"]), ("い", &["胃"])]);
    let response = server.candidate_response("あいi").await;
    assert_eq!(response, "4\n");
}

#[tokio::test]
async fn opcode1_okuri_ari_compound() {
    let user = vec![
        nashi("もんだい", &["問題"]),
        ari("なs", &["無", "済"]),
    ];
    let snap = build_snapshot(&user, &[]);
    let server = SkkServer::new(
        "test",
        "skkserv-compound",
        DictionaryStore::with_initial(snap),
        CompoundGeneratorConfig::default(),
    );
    let response = server.candidate_response("もんだいなs ").await;
    assert_eq!(response, "1/問題無/問題済/\n");
}

#[tokio::test]
async fn opcode1_ascii_input_does_not_extract_okuri() {
    let server = make_server(&[], &[("ka", &["カ"]), ("waii", &["ワイイ"])]);
    let response = server.candidate_response("kawaii ").await;
    assert_eq!(response, "1/カワイイ/\n");
}

#[tokio::test]
async fn opcode_zero_closes() {
    let server = make_server(&[], &[]);
    assert_eq!(
        server.handle_opcode('0', "", "127.0.0.1", 1178).await,
        OpcodeResult::Close
    );
}

#[tokio::test]
async fn opcode_two_returns_version() {
    let server = SkkServer::new(
        "v1",
        "skkserv-test",
        DictionaryStore::new(),
        CompoundGeneratorConfig::default(),
    );
    assert_eq!(
        server.handle_opcode('2', "", "127.0.0.1", 1178).await,
        OpcodeResult::Reply("skkserv-test/v1 ".to_string())
    );
}

#[tokio::test]
async fn opcode_three_returns_host_port() {
    let server = make_server(&[], &[]);
    let result = server.handle_opcode('3', "", "127.0.0.1", 1178).await;
    match result {
        OpcodeResult::Reply(body) => assert!(body.ends_with("/127.0.0.1:1178 "), "got: {}", body),
        _ => panic!("expected reply"),
    }
}

#[tokio::test]
async fn opcode_four_returns_four() {
    let server = make_server(&[], &[]);
    assert_eq!(
        server.handle_opcode('4', "なにか ", "127.0.0.1", 1178).await,
        OpcodeResult::Reply("4\n".to_string())
    );
}

#[tokio::test]
async fn opcode_one_integrates() {
    let server = make_server(&[], &[("あ", &["亜"]), ("い", &["胃"])]);
    assert_eq!(
        server.handle_opcode('1', "あい ", "127.0.0.1", 1178).await,
        OpcodeResult::Reply("1/亜胃/\n".to_string())
    );
}

#[tokio::test]
async fn unsupported_opcode_ignored() {
    let server = make_server(&[], &[]);
    assert_eq!(
        server.handle_opcode('9', "", "127.0.0.1", 1178).await,
        OpcodeResult::Ignore
    );
}

#[test]
fn decodes_euc_jp() {
    let text = "1あい ";
    let (cow, _enc, had_errors) = encoding_rs::EUC_JP.encode(text);
    assert!(!had_errors);
    let bytes: Vec<u8> = cow.into_owned();
    let decoded = IncomingCharset::EucJp.decode(&bytes).unwrap();
    assert_eq!(decoded, text);
    let operand: String = decoded.chars().skip(1).collect();
    let (body, okuri) = sanitize_yomi(&operand);
    assert_eq!(body, "あい");
    assert_eq!(okuri, None);
}

#[test]
fn extracts_multiple_requests_from_single_buffer() {
    let mut buffer: Vec<u8> = "1あい 1うえ ".as_bytes().to_vec();
    let messages = extract_messages(&mut buffer, IncomingCharset::Utf8);
    assert_eq!(messages.len(), 2);
    assert_eq!(messages[0].0, '1');
    assert_eq!(messages[0].1, "あい");
    assert_eq!(messages[1].0, '1');
    assert_eq!(messages[1].1, "うえ");
    assert!(buffer.is_empty());
}

#[test]
fn reassembles_across_buffers() {
    let full = "1あい ".as_bytes().to_vec();
    let split_at = full.len() / 2;
    let mut buffer: Vec<u8> = full[..split_at].to_vec();
    let first_pass = extract_messages(&mut buffer, IncomingCharset::Utf8);
    assert!(first_pass.is_empty());
    assert!(!buffer.is_empty());
    buffer.extend_from_slice(&full[split_at..]);
    let second_pass = extract_messages(&mut buffer, IncomingCharset::Utf8);
    assert_eq!(second_pass.len(), 1);
    assert_eq!(second_pass[0].0, '1');
    assert_eq!(second_pass[0].1, "あい");
    assert!(buffer.is_empty());
}

#[test]
fn recognizes_both_delimiters() {
    let mut buffer: Vec<u8> = "1あ\n1い ".as_bytes().to_vec();
    let messages = extract_messages(&mut buffer, IncomingCharset::Utf8);
    assert_eq!(messages.len(), 2);
    assert_eq!(messages[0].0, '1');
    assert_eq!(messages[0].1, "あ");
    assert_eq!(messages[1].0, '1');
    assert_eq!(messages[1].1, "い");
}

#[test]
fn keeps_unterminated_tail() {
    let mut buffer: Vec<u8> = "1あい 1うえ".as_bytes().to_vec();
    let messages = extract_messages(&mut buffer, IncomingCharset::Utf8);
    assert_eq!(messages.len(), 1);
    assert_eq!(messages[0].1, "あい");
    assert!(!buffer.is_empty());
}

#[test]
fn extracts_euc_jp_requests() {
    let text = "1あい 1うえ ";
    let (cow, _enc, had_errors) = encoding_rs::EUC_JP.encode(text);
    assert!(!had_errors);
    let mut buffer: Vec<u8> = cow.into_owned();
    let messages = extract_messages(&mut buffer, IncomingCharset::EucJp);
    assert_eq!(messages.len(), 2);
    assert_eq!(messages[0].1, "あい");
    assert_eq!(messages[1].1, "うえ");
}

#[test]
fn skips_empty_payloads() {
    let mut buffer: Vec<u8> = "  1あ ".as_bytes().to_vec();
    let messages = extract_messages(&mut buffer, IncomingCharset::Utf8);
    assert_eq!(messages.len(), 1);
    assert_eq!(messages[0].0, '1');
    assert_eq!(messages[0].1, "あ");
}

#[test]
fn sanitize_trims_and_strips() {
    let (body, okuri) = sanitize_yomi(" あい \n");
    assert_eq!(body, "あい");
    assert_eq!(okuri, None);

    let (body, okuri) = sanitize_yomi("abc");
    assert_eq!(body, "abc");
    assert_eq!(okuri, None);
}

#[test]
fn sanitize_extracts_okuri_prefix() {
    let (body, okuri) = sanitize_yomi("おくr");
    assert_eq!(body, "おく");
    assert_eq!(okuri, Some("r".to_string()));

    let (body, okuri) = sanitize_yomi("もんだいなs ");
    assert_eq!(body, "もんだいな");
    assert_eq!(okuri, Some("s".to_string()));
}

#[test]
fn sanitize_returns_empty_tuple() {
    let (body, okuri) = sanitize_yomi("   \n");
    assert_eq!(body, "");
    assert_eq!(okuri, None);
}
